File: answers/leetcode/Word Search/Word Search.java
class Solution {
    public boolean exist(char[][] board, String word) {
        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[i].length; j++) {
                if (this.bfs(i, j, board, 0, word)) {
                    return true;
                }
            }
        }
        return false;
    }

    private boolean bfs(int i, int j, char[][] board, int index, String word) {
        if (index >= word.length()) {
            return true;
        } else if (i < 0 || i >= board.length || j < 0 || j >= board[i].length || board[i][j] != word.charAt(index)) {
            return false;
        }
        board[i][j] = '.';
        if (this.bfs(i - 1, j, board, index + 1, word)
            || this.bfs(i + 1, j, board, index + 1, word)
            || this.bfs(i, j - 1, board, index + 1, word)
            || this.bfs(i, j + 1, board, index + 1, word)) {
            return true;
        } else {
            board[i][j] = word.charAt(index);
            return false;
        }
    }
}

File: answers/leetcode/Word Search/Word Search.py
#@result TLE

class Solution(object):
    def exist(self, board, word):
        """
        :type board: List[List[str]]
        :type word: str
        :rtype: bool
        """

        for i in range(len(board)):
            for j in range(len(board[i])):
                if self.dfs(i, j, board, 0, word):
                    return True
        return False

    def dfs(self, x, y, board, pos, word):
        if x < 0 or x >= len(board) or y < 0 or y >= len(board[x]) or '.' == board[x][y] or pos < 0:
            return False
        elif pos >= len(word):
            return True
        elif board[x][y] != word[pos]:
            return False
        board[x][y] = '.'
        if self.dfs(x - 1, y, board, pos + 1, word) or self.dfs(x + 1, y, board, pos + 1, word)or self.dfs(x, y - 1, board, pos + 1, word) or self.dfs(x, y + 1, board, pos + 1, word):
            return True
        board[x][y] = word[pos]
        return False
